content/posts/unpublished/ways-to-improve-be-api-performance/post.md

```diff
@@ -0,0 +1,304 @@
+a hybrid cache also incorporates a distributed caching component that spans multiple nodes or servers.
+
+## Pitfalls
+  * One of the most complex and challenging aspects of caching is deciding when and how to invalidate or update the cached data. Wrong **Caching invalidation** implementation might lead to losing advantages of caching or to providing outdated data to users;
+  * Cache consistency - during temporary unavailability of a cache certain updates of data occurs in a source, but not in the cache. After the cache becomes available again the changes must be synchronized. Another separate complex topic is consistency of a distributed cache nodes;
+  * Cache poisoning - caches are susceptible to various security threats, such as cache poisoning. Attackers inject malicious data into the cache compromising system integrity and user security, exposing harmful or fraudulent data to users.
+  * For hybrid caches like Infinispan upscaling or downscaling requires restarting all nodes. Infinispan always segments data that it stores in-memory and 
+changing the number of segments Infinispan creates requires a full cluster restart.
+
+# Asynchronous processing
+
+## Use cases and advantages
+
+In BE development, performance and responsiveness are intricately linked and play crucial role. Utilizing asynchronous processing for time-consuming or computation-intensive tasks is beneficial for both. For tasks such as email sending, files processing, and logging, where instant response is impossible or not necessary,
+asynchronous processing provides an efficient way to handle tasks without delaying other critical operations or blocking the main execution flow. Consider using of Java `CompletableFuture` and `@Async` Spring annotation for asynchronous processing.
+Let's demonstrate an example with email sending, using Spring `@Async` annotation. It allows to submit sending of email without waiting a response from actual email processing service.
+
+```java
+    @Async
+    public void sendEmail(EmailDetails emailDetails){
+        try {
+            EmailTemplate template = new EmailTemplate();
+            template.setFrom(emailDetails.getEmailSender);
+            template.setTo(emailDetails.getRecipients());
+            template.setText(emailDetails.getMessageBody());
+            template.setSubject(emailDetails.getSubject());
+            javaMailSender.send(template);
+            log.info("Mail sent successfully");
+        }catch (MailException exception){
+            log.debug("Error during sending an email", exception);
+        }
+    }
+```
+
+But `@Async` has its limitations; for large scale applications with more complex requirements and business logic using message queues like Apache Kafka, RabbitMQ or building an app using Event-Driven architecture is preferable.
```

Comment:
lets elaborate more here on the "why", for example, event are easier to manage since you have more control in non functional aspects such as throttling, partitioning, managing back-pressure, etc